README.md

```diff
@@ -4,222 +4,1008 @@
+Quite a big chunk of code this time, but when broken down it's pretty simple.
+
+We first define our `Config` struct, which is just your average POD struct with some data we want to persist.
+
+Then we use the `services.addFileConfig` function to create a `ServiceInfo` that describes an `AdaptableFileConfig`.
+
+The first template parameter is the user-defined type to store into the file, so `Config` in this case.
+
+The second template parameter is the adapter to use, which is the `AsdfConfigAdapter`.
+
+The first runtime parameter is the path to where the file should be stored.
+
+So we are using asdf to store/retrieve our `Config` from the file `"./config.json"`, in essence.
+
+We can access this service from our commands by requesting an `IConfig!Config`.
+
+Over inside of the `SeedConfigCommand` struct, we have an instance of `IConfig!Config` injected, and then inside of `onExecute` we do something
+a bit more peculiar.
+
+The `IConfig.edit` function is used here, as I want to demonstrate the handful of capabilities that `IConfig` supports. As the comment says, there's
+a shortcut function for this particular usage called `editAndSave`.
+
+The first parameter is a delegate (lambda) that is given a shallow copy of the configuration's value by reference. All this delegate needs to do is
+populate the configuration value with whatever it wants to set it to, by whatever means to get the data.
+
+The second parameter is a flag called `RollbackOnFailure`. As the name implies, if the delegate in the first parameter throws an exception then the
+`IConfig` will attempt to rollback any changes. Please see [this comment](https://jcli.dpldocs.info/jaster.cli.config.IConfig.edit.html) on it works exactly.
+
+The third and final parameter is a flag called `SaveOnSuccess`, which literally does at it says on the tin. If the delegate was successful, then call
+`IConfig.save` to save any changes.
+
+Finally with this `onExecute`, we just make sure that the return value was `WasExceptionThrown.no`, which explains itself.
+
+Alternatively you can just set `IConfig.value` to something, and then call `IConfig.save`. The `IConfig.edit` function was just supposed to be a helper
+around things.
+
+We're now onto the final part of this code which is the `PrintConfigCommand`.
+
+Literally all it does it ask for the config to be injected, retrieves its value via `IConfig.value`, and then prints it to the screen
+(D's `writeln` automatically pretty prints structs).
+
+As I said, JCLI's built-in configuration isn't terribly fancy, and offloads the majority of the work onto third party code. But it gets the job
+done when you just want something up quick and easy.
+
+## Inheritance
+
+JCLI supports command inheritance.
+
+The only rules with inheritance are:
+
+* Only concrete classes can be marked with `@Command`.
+
+* Concrete classes must have `onExecute` defined, either by a base class or directly.
```

Comment:
This is mostly just an issue with terminology, and my apparent lack of it.

You could have `abstract class CommandBase` which is a non-concrete class (since its abstract).

Then `class CommandA : CommandBase` which is a concrete class (but isn't final).

Then `final class CommandB : CommandA` which is also a concrete class (but is final).

I think I'm just missing the proper word for "concrete".

Reply:
I don't really know *why* you'd make a fully-fledged command that inherits another fully-fledged, non-abstract command, but it's possible.

Reply:
I couldn't think of any inheritance usage at all but it's up to you.

Reply:
I remember when I was checking out dub's code a while ago (2 years I think) that the commands in dub used inheritance.

Personally, I wouldn't really use it myself, but it seems there's a precedent for it, which is why it's supported.